system/jlib/jdebug.cpp

```diff
@@ -1860,33 +1860,44 @@ void OsDiskInfo::initMajorMinor()
     Owned<IPipeProcess> pipe = createPipeProcess();
     if (pipe->run("list disks", cmd, nullptr, false, true, true, 8192))
     {
-        StringBuffer output;
-        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
-        readSimpleStream(output, *pipeReader);
-        unsigned exitcode = pipe->wait();
-        if ( (exitcode == 0) && (output.length() > 0) )
+        bool timedout = false;
+        // handle when lsblk runs too long
+        unsigned exitcode = pipe->wait(2000, timedout);
+        if ( (!timedout) && (exitcode == 0) )
         {
-            StringArray lines;
-            lines.appendList(output, "\n");
-            ForEachItemIn(idx, lines)
+            StringBuffer output;
+            Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+            readSimpleStream(output, *pipeReader);
+            if (output.length() > 0)
             {
-                // line: TYPE="disk" MAJ:MIN="259:0"
-                unsigned majnum, minnum;
-                if (2 == sscanf(lines.item(idx), "TYPE=\"disk\" MAJ:MIN=\"%u:%u\"", &majnum, &minnum))
+                StringArray lines;
+                lines.appendList(output, "\n");
+                ForEachItemIn(idx, lines)
                 {
-                    unsigned mm = (majnum<<16)+minnum;
-                    diskMajorMinor.appendUniq(mm);
+                    // line: TYPE="disk" MAJ:MIN="259:0"
+                    unsigned majnum, minnum;
+                    if (2 == sscanf(lines.item(idx), "TYPE=\"disk\" MAJ:MIN=\"%u:%u\"", &majnum, &minnum))
+                    {
+                        unsigned mm = (majnum<<16)+minnum;
+                        diskMajorMinor.appendUniq(mm);
+                    }
                 }
+                return;
             }
         }
-        else
-        {
-            StringBuffer outputErr;
-            Owned<ISimpleReadStream> pipeReaderErr = pipe->getErrorStream();
-            readSimpleStream(outputErr, *pipeReaderErr);
-            if (outputErr.length() > 0)
-                WARNLOG("WARNING: Pipe: output: %s", outputErr.str());
-        }
+        if (timedout)
+            pipe->abort();
+        // log unexpected output, if any
+        StringBuffer output;
+        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+        readSimpleStream(output, *pipeReader);
```

Comment:
Yes, I should add an additional condition, thanks.
I will common up some code also.

Reply:
But without changing createPipeProcess() we must do the wait() before any readSimpleStream().

Reply:
without changing createPipeProcss() you could do the reading (readSimpleStream etc.) on it's own thread.
If the pipe can block (on write), then I'm not sure there's any alternative to doing it async either by your own thread handling, or thread handling in the pipe impl.

Reply:
```
    forkthread.setown(new cForkThread(this));
```
called from run() creates a new thread which calls the waitpid().
wait(timeoutms) waits the specified amount for that thread to end, if it does not then we can kill the child process in abort().
As long as the wait(timeoutms) is called before read() it will not block no matter how much output is produced.

Reply:
I wasn't suggesting it blocked indefinitely.
I mean it the child process will block because the pipe buffer is full,
pipe->wait() will be blocked waiting on the child pid *until* it hits the wait timeout.

This is even if the child program would have finished gracefully after e.g 0.1 secs.
i.e. if the child process writes 65K quickly, it will cause this block, timeout and treat it as if the child process failed. So really that's not correct in general it should handle lsblk outputting 65K I would have thought.

Now, can argue that lsblk is never going to write >64K and finish, so it's ok to fail over like this?

Reply:
See comments above.